libs/domains/environment/src/lib/slices/deployment-stage.ts

```diff
@@ -1,203 +1,212 @@
+    }) => {
+      const response = await deploymentStageMainCallApi.attachServiceToDeploymentStage(deploymentStageId, serviceId)
+      return response.data.results
+    },
+    {
+      onSuccess: (_, variables) => {
+        queryClient.invalidateQueries(['environment', environmentId, 'deploymentStageList'])
+        if (variables.prevStage) {
+          // default toast when we don't apply undo
+          toast(
+            ToastEnum.SUCCESS,
+            'Your deployment stage is updated',
+            'Do you need to go back?',
+            () =>
+              currentUseMutation.mutate({
+                deploymentStageId: variables.prevStage?.deploymentStageId || ' ',
+                serviceId: variables.prevStage?.serviceId || '',
+              }),
+            '',
+            'Undo'
+          )
+        } else {
+          // default toast when we apply undo
+          toast(ToastEnum.SUCCESS, 'Your deployment stage is updated')
+        }
+      },
+      onError: (err) => toastError(err as SerializedError),
+    }
+  )
+
+  return currentUseMutation
+}
+
+export const useMoveDeploymentStageRequested = (onSuccessCallback: (result: DeploymentStageResponse[]) => void) => {
+  const queryClient = useQueryClient()
+
+  return useMutation(
+    async ({
+      stageId,
+      beforeOrAfterStageId,
+      after,
+    }: {
+      stageId: string
+      beforeOrAfterStageId: string
+      after: boolean
+    }) => {
+      let response
+      if (after) {
+        response = await deploymentStageMainCallApi.moveAfterDeploymentStage(stageId, beforeOrAfterStageId)
+      } else {
+        response = await deploymentStageMainCallApi.moveBeforeDeploymentStage(stageId, beforeOrAfterStageId)
       }
-      environmentsAdapter.updateOne(state, update)
-      toast(ToastEnum.SUCCESS, 'Your stage order has been successfully updated')
-    })
-    .addCase(moveDeploymentStageRequested.rejected, (state: EnvironmentsState, action) => {
-      toastError(action.error)
-    })
+      return response.data.results
+    },
+    {
+      onSuccess: (data) => {
+        if (data) {
+          queryClient.setQueryData<DeploymentStageResponse[] | undefined>(
+            ['environment', data[0].environment.id, 'deploymentStageList'],
+            data
+          )
+          toast(ToastEnum.SUCCESS, 'Your stage order has been successfully updated')
+          onSuccessCallback(data)
+        }
+      },
+      onError: (err: SerializedError) => toastError(err),
+    }
+  )
+}
+
+export const useCreateEnvironmentDeploymentStage = (
+  environmentId: string,
+  onSuccessCallback: () => void,
+  onSettledCallback: () => void
+) => {
+  const queryClient = useQueryClient()
+
+  return useMutation(
+    async ({ data }: { data: DeploymentStageRequest }) => {
+      const response = await deploymentStageMainCallApi.createEnvironmentDeploymentStage(environmentId, data)
+      return response.data
+    },
+    {
+      onMutate: async (variables) => {
```

Comment:
Not necessary to do it but :
- This function will fire before the mutation function is fired and is passed the same variables the mutation function would receive
- Useful to perform optimistic updates to a resource in hopes that the mutation succeeds
- The value returned from this function will be passed to both the onError and onSettled functions in the event of a mutation failure and can be useful for rolling back optimistic updates.